File: kotlin/1361-validate-binary-tree-nodes.kt
class Solution {
    fun validateBinaryTreeNodes(n: Int, leftChild: IntArray, rightChild: IntArray): Boolean {
        val hasParent = hashSetOf<Int>().apply {
            for (n in (leftChild + rightChild)) {
                if (n != -1)
                    this.add(n)
            }
        }

        if (hasParent.size == n) return false

        var root = -1
        for (i in 0 until n) {
            if (i !in hasParent) {
                root = i
                break
            }
        }

        val visit = hashSetOf<Int>()
        fun dfs(i: Int): Boolean {
            if (i == -1) return true
            if (i in visit) return false  

            visit.add(i)

            return dfs(leftChild[i]) && dfs(rightChild[i])
        }

        return dfs(root) && visit.size == n
    }
}

File: javascript/1361-validate-binary-tree-nodes.js
/**
 * DFS
 * Time O(n) | Space O(n)
 * https://leetcode.com/problems/validate-binary-tree-nodes/
 * 
 * @param {number} n
 * @param {number[]} leftChild
 * @param {number[]} rightChild
 * @return {boolean}
 */
var validateBinaryTreeNodes = function(n, leftChild, rightChild) {
    
    const visited = new Set();

    const findRoot = () => {
      
      const childrenSet = new Set();
      for (let i = 0; i < n; i++) {
        childrenSet.add(i);
      }
      
      for (let i = 0; i < n; i++) {
        childrenSet.delete(leftChild[i]);
        childrenSet.delete(rightChild[i]);
      }

      return [...childrenSet][0];
    }

    const dfs = (i) => {

      if (i === -1) return true;
      if (visited.has(i)) return false;

      const left = leftChild[i];
      const right = rightChild[i];
      visited.add(i);
      return dfs(left) && dfs(right);
    }

    const root = findRoot();
    return dfs(root) && visited.size === n;
};
